### DISCUSSION TITLE
Интеграция с Amocrm

### DISCUSSION THREAD

**Discussion by Inkerson** [totumonline/totum-ru-issues-and-discussions#169] [Как это сделать?] [answered]
Есть ли возможность связки с амосрм без использования сторонних сервисов. Задача получать и отправлять статусы по сделкам.

**Comment by erpic-ru** [comment]
Да, можно попробовать сделать интеграцию через вебхуки от амоцрм и настройку Remotes в Тотуме
https://www.amocrm.ru/developers/content/crm_platform/webhooks-format
https://ru.docs.totum.online/api#type_remotes

**Reply by ttm-alexeypolunin** [reply]
Я посмотрел документацию АМО — Тимур предлагает норм вариант, если нужно из АМО отправлять в Тотум.

Если из Тотума надо отправлть в АМО, то вот эта часть в их документации:

https://www.amocrm.ru/developers/content/crm_platform/leads-api#leads-edit это редактирование сделки,

там же есть примеры на создание https://www.amocrm.ru/developers/content/crm_platform/leads-api#leads-add

Со стороны АМО:

```
Предусмотрены механизмы ограничения активности работы с API – не более 7 запросов в секунду, а также в некоторых методах предусмотрены ограничения на количества возвращаемых за один запрос данных (подробнее в описании конкретных методов).
```

```
Все запросы необходимо производить не на общий домен http://www.amocrm.ru, а на точный адрес вашего аккаунта, например, https://company.amocrm.ru
```

### Для авторизации надо получить токен доступа и потом передавать его в заголовке запроса:

Вот про это https://www.amocrm.ru/developers/content/oauth/step-by-step#%D0%97%D0%B0%D0%BF%D1%80%D0%BE%D1%81%D1%8B-%D0%BA-API-%D1%81-%D0%BF%D0%B5%D1%80%D0%B5%D0%B4%D0%B0%D1%87%D0%B5%D0%B9-access-token

Вроде как можно сделать проще, получив доступ от администратора конкретного аккаунта: https://www.amocrm.ru/developers/content/oauth/easy-auth

> Я эти ссылки оставляю здесь, что бы наш программист посмотрел, как у него будет время и сделал пример запроса из Тотум

**Reply by Inkerson** [reply]
спасибо

**Reply by Inkerson** [reply]
Если планируете дизайн переработать, обращайтесь, денег не возьму.

**Reply by ttm-alexeypolunin** [reply]
Мы сделаем настройку цветовой схемы, но базовые принципы такие:

— монохромность
— теплые тона
— совместимость с палитрой web_color
— плотность
— отсутсвие глянца
— отсутсвие анимаций

И в чем здесь сложность — реально миллион экранов, что бы все было ровно при перерисовке, надо год всадить (это уже проверяли, чуваки рисовали для своего проекта кастом-дизайн — основные экраны были норм но недолго, а мелочи так сразу были покосившиеся).

**Reply by ttm-alexeypolunin** [reply]
### Получаем данные из amo через web_hooks

Создаем `Простую таблицу` + `Поле` тип `Данные` в `Cтрочной части`. Они как угодно могут называть — это отладочная таблица.

Создаем `Remout` в `Системные таблицы` — `Основные` — `API` — `Remotes`.

В пользователях Системные таблицы — Доступы — Пользователи создаем пользователя `amo` с типом `api` (пароль и прочие настройки можно не указывать).

Называем его с использованием хэша, например: `amo_leads_add_RhKsDuJf5YUakHd4`. Вы хэш только свой придумайте пжлст :)

Для добавления, изменения, удаления сделки лучше добавить три разных ремоута и сделать три разных хука. Пока делаем один для добавления.

Выбираем в `Remote user` пользователя `amo`.

В код пишем (замените `OUR_TABLE_NAME` и `FIELD_NAME` на созданные вами ранее): 

```
=: insert(table: 'OUR_TABLE_NAME'; field: 'FIELD_NAME' = $#post)
```

В `amo` идем в раздел `amoМаркет`.

Вверху справа нажимаем `+ WEB HOOKS`.

В адрес пишем `https://YOU_HOST/Remotes/YOU_REMOTE_NAME` и ставим галочки напротив `Добавление сделки` и жмем `Сохранить`. Напоминаю, что на добавление, изменение и удаление лучше сделать три разных хука. Начинаем с добавления.

В моем случае это: `https://test.ttmapp.ru/Remotes/amo_leads_add_RhKsDuJf5YUakHd4` 

Проверяем, добавляем сделку в `amo`. Смотрим, что добавилось нам в отладочную таблицу таблицу. Что-то такое:

```
{
  "leads": {
    "add": [
      {
        "id": "2413525",
        "name": "Сделка 1",
        "price": "10000",
        "status_id": "49613212",
        "account_id": "30305773",
        "created_at": "1659107270",
        "updated_at": "1659107270",
        "date_create": "1659107270",
        "pipeline_id": "5630533",
        "last_modified": "1659107270",
        "created_user_id": "8408266",
        "modified_user_id": "8408266",
        "responsible_user_id": "8408266"
      }
    ]
  },
  "account": {
    "id": "30305773",
    "_links": {
      "self": "https://alexeypolunin.amocrm.ru"
    },
    "subdomain": "alexeypolunin"
  }
}
```

Создаем в тотум таблицу со сделками и модифицируем код, добавляющий сделку в Remotes:

```
=: insertList(table: 'sdelki'; field: 'id_sdelki' = $#post[leads][add][[id]]; field: 'nazvanie_sdelki' = $#post[leads][add][[name]]; log: true)
```

Проверяем.

Добавляем теперь второй ремоут на изменение: `amo_leads_edit_RhKsDuJf5YUakHd4`

Все тоже самое, сначала направляем его в отладочную таблицу:
```
=: insert(table: 'OUR_TABLE_NAME'; field: 'FIELD_NAME' = $#post)
```

Аналогично добавляем второй хук в `amo`. Проверяем. Смотрим в таблицу отладки. Пришло что-то типа:

```
{
  "leads": {
    "update": [
      {
        "id": "2413657",
        "name": "Сделка edit",
        "price": "15000",
        "status_id": "49613212",
        "account_id": "30305773",
        "created_at": "1659107889",
        "updated_at": "1659108303",
        "date_create": "1659107889",
        "pipeline_id": "5630533",
        "last_modified": "1659108303",
        "created_user_id": "8408266",
        "modified_user_id": "8408266",
        "responsible_user_id": "8408266"
      }
    ]
  },
  "account": {
    "id": "30305773",
    "_links": {
      "self": "https://alexeypolunin.amocrm.ru"
    },
    "subdomain": "alexeypolunin"
  }
}
```

Модифицируем код во втором ремоуте:

```
=: setListExtended(table: 'sdelki'; field: 'nazvanie_sdelki' = $#post[leads][update][[name]]; where: 'id_sdelki' = $#post[leads][update][[id]]; log: true)
```

И третий все тоже самое для удаления, приходит:

```
{
  "leads": {
    "delete": [
      {
        "id": "2413981",
        "status_id": "49613212",
        "pipeline_id": "5630533"
      }
    ]
  },
  "account": {
    "id": "30305773",
    "_links": {
      "self": "https://alexeypolunin.amocrm.ru"
    },
    "subdomain": "alexeypolunin"
  }
```

Модифицируем код ремоута удаления:

```
=: deleteList(table: 'sdelki'; where: 'id_sdelki' = $#post[leads][delete][[id]]; log: true)
```

### Замечание по Тотум

Скорее всего у вас `Сделки` — это таблица циклов. И параметры, которые вы будете получать приведенными выше кодами будут приходит в таблицу `Циклов`, а не внутрь циклов. Поэтому с данными внутри цикла их надо связать через `set` и `clear`: https://ru.docs.totum.online/label-type-fields

Также вы можете сконструировать ссылку в Тотум на сделку в `amo` используя тип поля Строка, отображая его как `url` или кнопку с функцие `linkToScript`.  

**Создание обратной интеграции из `Тотум` в `amo` в следующей серии.**

**Comment by ttm-alexeypolunin** [comment]
Поставил на Четверг 28.07.22 посмотреть доки по АМО. В одном из проектов у партнеров была синхронизация — но мне нужно время, что бы пример кода сделать.

**Reply by Inkerson** [reply]
Спасибо, будем ждать

**Reply by ttm-alexeypolunin** [reply]
Если у вас там более-менее законченный кейс получится — напишите мне в telegram, было бы здорово его в рассылку amo запихнуть.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### Сделали интеграцию: https://ru.docs.totum.online/totum-amo

Видео: https://www.youtube.com/watch?v=XgFHB8Fyq7U

Файл семпла для загрузки: https://raw.githubusercontent.com/totumonline/ttm-samples/main/samples/ttm_amo.json.gz.ttm

Инструкция по импорту Семпла: https://youtu.be/UA7j28Ht5u4
